csharp/src/IceLocatorDiscovery/PluginI.cs

```diff
@@ -256,350 +173,224 @@ public void
 
-                        locator.Communicator.Logger.Warning(
-                            "received Ice locator with different instance name:\n" +
-                            "using = `" + _locator.Identity.Category + "'\n" +
-                            "received = `" + locator.Identity.Category + "'\n" +
-                            "This is typically the case if multiple Ice locators with different " +
-                            "instance names are deployed and the property `IceLocatorDiscovery.InstanceName'" +
-                            "is not set.");
+                        var s = new StringBuilder();
+                        s.Append("received Ice locator with different instance name:\n")
+                         .Append("using = `").Append(_locator.Identity.Category).Append("'\n")
+                         .Append("received = `").Append(locator.Identity.Category).Append("'\n")
+                         .Append("This is typically the case if multiple Ice locators with different ")
+                         .Append("instance names are deployed and the property `IceLocatorDiscovery.InstanceName' ")
+                         .Append("is not set.");
+                        locator.Communicator.Logger.Warning(s.ToString());
                     }
                     return;
                 }
 
-                if (_pending) // No need to continue, we found a locator
-                {
-                    _timer.Cancel(this);
-                    _pendingRetryCount = 0;
-                    _pending = false;
-                }
-
                 if (_traceLevel > 0)
                 {
                     var s = new StringBuilder("locator lookup succeeded:\nlocator = ");
                     s.Append(locator);
-                    if (_instanceName.Length == 0)
+                    if (_instanceName.Length > 0)
                     {
                         s.Append("\ninstance name = ").Append(_instanceName);
                     }
                     _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
                 }
 
-                ILocatorPrx? l = null;
-                if (_pendingRequests.Count == 0)
+                if (_locator == null)
                 {
-                    _locators.TryGetValue(locator.Identity.Category, out _locator);
+                    if (_completionSource == null)
+                    {
+                        // Ignore reply to request that already timeout.
+                        return;
+                    }
+                    _locator = locator;
+                    if (_instanceName.Length == 0)
+                    {
+                        _instanceName = _locator.Identity.Category; // Stick to the first locator
+                    }
+                    Debug.Assert(_completionSource.Task.IsCompleted == false);
+                    _completionSource.SetResult(locator);
+                    _completionSource = null;
```

Comment:
I would just use `_completionSource?.TrySetResult(locator);` and leave the responsibility of clearing this data member to `FindLocatorAsync` once the await returns. Same for clearing _findLocatorTask, it would be clearer to do this in `FindLocatorAsync`.

Reply:
Fine we will have to add a check for IsCompleted we don't want to call SetResult on a completed task